sherpa/astro/io/tests/test_backend_switch.py

```diff
@@ -0,0 +1,50 @@
+#
+#  Copyright (C) 2021
+#  MIT
+#
+#
+#  This program is free software; you can redistribute it and/or modify
+#  it under the terms of the GNU General Public License as published by
+#  the Free Software Foundation; either version 3 of the License, or
+#  (at your option) any later version.
+#
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License along
+#  with this program; if not, write to the Free Software Foundation, Inc.,
+#  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+import pytest
+
+from sherpa.utils.testing import requires_data, requires_fits
+from sherpa.astro import io
+from sherpa.astro import ui
+
+
+@requires_fits
+@requires_data
+def test_backend_switch(make_data_path):
+    '''Test that we can switch backends.
+
+    Of course, switching only makes sense if there is more than one backend
+    to try, so this is listed as "requires_fits".
+    '''
+    pha = make_data_path("3c273.pi")
+    ui.load_pha(pha)
+
+    orig_backend = io.backend
+
+    import sherpa.astro.io.dummy_backend
+    io.backend = sherpa.astro.io.dummy_backend
+
+    with pytest.raises(NotImplementedError,
+                       match="No usable I/O backend was imported."):
+        pha = make_data_path("3c273.pi")
+        ui.load_pha(pha)
+
+    io.backend = orig_backend
```

Comment:
Technically the test should be something like

```python

pha = mane_data_path('3c273.pi')
ui.load_pha(pha)

orig_backend = io.backend
try:
    import sherpa.astro.io.dummy_backend
    io.backend = sherpa.astro.io.dummy_backend

    with pytest.raises(NotImplementedError,
                       match="No usable I/O backend was imported."):
        ui.load_pha(pha)

finally:
    io.backend = orig_backend

ui.load_pha(pha)
```

just in case there's any problem when switching the backend so we don't ruin other tests, but I'm not sure it's worth it/doesn't cause more problems than is stops

Reply:
It's a test. If it fails something is wrong and the backend can't be switched back, then that needs to be fixed. In principle, I don't think what you suggest is the right idea, but I think the way I've written it is easier. If it works, it's the same. If it fails it needs to be fixed, finally clause or not.